src/ivector/xvector.cc

```diff
@@ -0,0 +1,140 @@
+// ivector/xvector.cc
+
+// Copyright 2016     Daniel Povey
+//                    David Snyder
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ivector/xvector.h"
+
+namespace kaldi {
+
+void ComputeXvectorObjfAndDeriv(
+    const CuMatrixBase<BaseFloat> &xvector_pairs,
+    const CuSpMatrix<BaseFloat> &S,
+    BaseFloat b, CuMatrixBase<BaseFloat> *deriv_xvector,
+    CuVector<BaseFloat> *deriv_S_and_b, BaseFloat *tot_objf,
+    BaseFloat *tot_weight) {
+
+  BaseFloat same_objf = 0,
+            diff_objf = 0;
+  int32 S_dim = S.NumCols() * (S.NumCols() + 1) / 2,
+        N = xvector_pairs.NumRows(),
+        xvector_dim = xvector_pairs.NumCols();
+  BaseFloat K = 1.0 / (N - 2.0);
+
+  if (deriv_xvector == NULL)
+    KALDI_ASSERT(deriv_S_and_b == NULL);
+  else {
+    KALDI_ASSERT(deriv_xvector->NumCols() == xvector_dim);
+    KALDI_ASSERT(deriv_xvector->NumRows() == N);
+    KALDI_ASSERT(deriv_S_and_b->Dim() == S_dim + 1);
+  }
+
+  // Handle portion of the objf corresponding to pairs of xvectors
+  // from the same classes.
+  for (int32 i = 0; i < N/2; i++) {
```

Comment:
This looks plausible from the point of view of the math, but it's very against the spirit of CUDA programming to have loops like this.  The idea is to have all the parallel stuff happening on the GPU card, so you go to great lengths to do things in as few matrix operations as possible.   But I think it's best if you push on with the rest of this project and let me think about the best way to do this- I'll let you know when I've figured out an efficient way to do it.